flink-table/flink-sql-parser/src/test/java/org/apache/flink/sql/parser/FlinkSqlParserImplTest.java

```diff
@@ -0,0 +1,447 @@
+	}
+
+	@Test
+	public void testCreateTableWithWatermarkBoundedDelay3() {
+		check("CREATE TABLE tbl1 (\n" +
+				"  a bigint,\n" +
+				"  b varchar, \n" +
+				"  c as 2 * (a + 1), \n" +
+				"  WATERMARK wk FOR a AS BOUNDED WITH DELAY 1000 SECOND\n" +
+				")\n" +
+				"  with (\n" +
+				"    connector = 'kafka', \n" +
+				"    kafka.topic = 'log.test'\n" +
+				")\n",
+			"CREATE TABLE `tbl1` (\n" +
+				"  `a`  BIGINT,\n" +
+				"  `b`  VARCHAR,\n" +
+				"  `c` AS (2 * (`a` + 1)),\n" +
+				"  WATERMARK `wk` FOR `a` AS BOUNDED WITH DELAY 1000 SECOND\n" +
+				") WITH (\n" +
+				"  `connector` = 'kafka',\n" +
+				"  `kafka`.`topic` = 'log.test'\n" +
+				")");
+	}
+
+	@Test
+	public void testCreateTableWithNegativeWatermarkOffsetDelay() {
+		checkFails("CREATE TABLE tbl1 (\n" +
+				"  a bigint,\n" +
+				"  b varchar, \n" +
+				"  c as 2 * (a + 1), \n" +
+				"  WATERMARK wk FOR a AS BOUNDED WITH DELAY ^-^1000 SECOND\n" +
+				")\n" +
+				"  with (\n" +
+				"    connector = 'kafka', \n" +
+				"    kafka.topic = 'log.test'\n" +
+				")\n",
+			"(?s).*Encountered \"-\" at line 5, column 44.\n" +
+				"Was expecting:\n" +
+				"    <UNSIGNED_INTEGER_LITERAL> ...\n" +
+				".*");
+	}
+
+	@Test
+	public void testCreateTableWithWatermarkStrategyAscending() {
+		check("CREATE TABLE tbl1 (\n" +
+				"  a bigint,\n" +
+				"  b varchar, \n" +
+				"  c as 2 * (a + 1), \n" +
+				"  WATERMARK wk FOR a AS ASCENDING\n" +
+				")\n" +
+				"  with (\n" +
+				"    connector = 'kafka', \n" +
+				"    kafka.topic = 'log.test'\n" +
+				")\n",
+			"CREATE TABLE `tbl1` (\n" +
+				"  `a`  BIGINT,\n" +
+				"  `b`  VARCHAR,\n" +
+				"  `c` AS (2 * (`a` + 1)),\n" +
+				"  WATERMARK `wk` FOR `a` AS ASCENDING\n" +
+				") WITH (\n" +
+				"  `connector` = 'kafka',\n" +
+				"  `kafka`.`topic` = 'log.test'\n" +
+				")");
+	}
+
+	@Test
+	public void testCreateTableWithWatermarkStrategyFromSource() {
+		check("CREATE TABLE tbl1 (\n" +
+				"  a bigint,\n" +
+				"  b varchar, \n" +
+				"  c as 2 * (a + 1), \n" +
+				"  WATERMARK wk FOR a AS FROM_SOURCE\n" +
+				")\n" +
+				"  with (\n" +
+				"    connector = 'kafka', \n" +
+				"    kafka.topic = 'log.test'\n" +
+				")\n",
+			"CREATE TABLE `tbl1` (\n" +
+				"  `a`  BIGINT,\n" +
+				"  `b`  VARCHAR,\n" +
+				"  `c` AS (2 * (`a` + 1)),\n" +
+				"  WATERMARK `wk` FOR `a` AS FROM_SOURCE\n" +
+				") WITH (\n" +
+				"  `connector` = 'kafka',\n" +
+				"  `kafka`.`topic` = 'log.test'\n" +
+				")");
+	}
+
+	@Test
+	public void testCreateTableWithComplexType() {
+		check("CREATE TABLE tbl1 (\n" +
+			"  a ARRAY<bigint>, \n" +
+			"  b MAP<int, varchar>,\n" +
+			"  c STRUCT<cc0:int, cc1: float, cc2: varchar>,\n" +
+			"  PRIMARY KEY (a, b) \n" +
+			") with (\n" +
+			"  x = 'y', \n" +
+			"  asd = 'data'\n" +
+			")\n", "CREATE TABLE `tbl1` (\n" +
+			"  `a`  ARRAY< BIGINT >,\n" +
+			"  `b`  MAP< INTEGER, VARCHAR >,\n" +
+			"  `c`  STRUCT< `cc0` : INTEGER, `cc1` : FLOAT, `cc2` : VARCHAR >,\n" +
```

Comment:
Does we support nested complex type? like `ARRAY<ARRAY<bigint>>`, add some cases?

Reply:
I agree, added